troubadix/reporter.py

```diff
@@ -0,0 +1,205 @@
+
+from pathlib import Path
+from typing import List
+from pontos.terminal.terminal import Terminal
+from troubadix.helper.helper import get_path_from_root
+
+from troubadix.plugin import (
+    LinterError,
+    LinterFix,
+    LinterMessage,
+    LinterResult,
+    LinterWarning,
+)
+from troubadix.results import FileResults, ResultCounts
+
+
+class Reporter:
+    def __init__(
+        self,
+        term: Terminal,
+        root: Path,
+        *,
+        fix: bool = False,
+        log_file: Path = None,
+        statistic: bool = True,
+        verbose: int = 0,
+    ) -> None:
+        self._term = term
+        self._log_file = log_file
+        self._statistic = statistic
+        self._verbose = verbose
+        self._fix = fix
+        self._files_count = 0
+        self._root = root
+
+        self._result_counts = ResultCounts()
+
+    def set_files_count(self, count: int):
+        self._files_count = count
+
+    def get_result_count(self) -> int:
+        return self._result_counts.error_count
+
+    def _report_warning(self, message: str) -> None:
+        self._term.warning(message)
+        self._log_append(f"\t\t{message}".replace("\n", "\n\t\t"))
+
+    def _report_error(self, message: str) -> None:
+        self._term.error(message)
+        self._log_append(f"\t\t{message}".replace("\n", "\n\t\t"))
+
+    def report_info(self, message: str) -> None:
+        """Report an info message"""
+        self._term.info(message)
+        self._log_append(f"\t{message}")
+
+    def _report_bold_info(self, message: str) -> None:
+        self._term.bold_info(message)
+        self._log_append(f"\n\n{message}")
+
+    def _report_ok(self, message: str) -> None:
+        self._term.ok(message)
+        self._log_append(f"\t\t{message}".replace("\n", "\n\t\t"))
+
+    def _process_plugin_results(
+        self, plugin_name: str, plugin_results: List[LinterResult]
+    ):
+        """Process the results of a plugin: Print/Log results if
+        verbosity/logging fits and count the results"""
+        if plugin_results and self._verbose > 0:
+            self.report_info(f"Results for plugin {plugin_name}")
+        elif self._verbose > 2:
+            self._report_ok(f"No results for plugin {plugin_name}")
+
+        # add the results to the statistic and print/log them
+        with self._term.indent():
+            for plugin_result in plugin_results:
+                if isinstance(plugin_result, LinterError):
+                    self._result_counts.add_error(plugin_name)
+                    report = self._report_error
+                elif isinstance(plugin_result, LinterWarning):
+                    self._result_counts.add_warning(plugin_name)
+                    report = self._report_warning
+                elif isinstance(plugin_result, LinterFix):
+                    self._result_counts.add_fix(plugin_name)
+                    report = self._report_ok
+                elif isinstance(plugin_result, LinterMessage):
```

Comment:
I think we should fix the class hierarchy and remove LinterMessage because it is unused.
```suggestion
                elif isinstance(plugin_result, LinterResult):
```